beacon-chain/db/kv/state.go

```diff
@@ -197,16 +286,71 @@ func (s *Store) DeleteStates(ctx context.Context, blockRoots [][32]byte) error {
 	return nil
 }
 
-// creates state from marshaled proto state bytes.
-func createState(ctx context.Context, enc []byte) (*pb.BeaconState, error) {
+// creates state from marshaled proto state bytes. Also add the validator entries retrieved
+// from the validator bucket and complete the state construction.
+func createState(ctx context.Context, enc []byte, validatorEntries []*ethpb.Validator) (*pb.BeaconState, error) {
 	protoState := &pb.BeaconState{}
 	if err := decode(ctx, enc, protoState); err != nil {
 		return nil, errors.Wrap(err, "failed to unmarshal encoding")
 	}
+	if validatorEntries == nil {
+		validatorEntries = make([]*ethpb.Validator, 0)
+	}
+	protoState.Validators = validatorEntries
 	return protoState, nil
 }
 
-// HasState checks if a state by root exists in the db.
+// retrieve the validator entries for a given bock root. These entries are stored in a
+// separate bucked to save state.
+func (s *Store) validatorEntries(ctx context.Context, blockRoot [32]byte) ([]*ethpb.Validator, error) {
+	ctx, span := trace.StartSpan(ctx, "BeaconDB.validaotrEntries")
+	defer span.End()
+	var validatorEntries []*ethpb.Validator
+	err := s.db.View(func(tx *bolt.Tx) error {
+		// get the validator keys from the index bucket
+		idxBkt := tx.Bucket(BlockRootValidatorKeysIndexBucket)
+		valKey := idxBkt.Get(blockRoot[:])
+		if len(valKey) == 0 {
+			return errors.Errorf("invalid compressed validator keys length")
+		}
+
+		// decompress the keys and check if they are of proper length.
+		validatorKeys, sErr := snappy.Decode(nil, valKey)
+		if sErr != nil {
+			return errors.Wrap(sErr, "failed to uncompress validator keys")
+		}
+		if len(validatorKeys)%hashLength != 0 {
+			return errors.Errorf("invalid validator keys length: %d", len(validatorKeys))
+		}
+
+		// get the corresponding validator entries from the validator bucket.
+		valBkt := tx.Bucket(stateValidatorsBucket)
+		for i := 0; i < len(validatorKeys); i += hashLength {
+			key := validatorKeys[i : i+hashLength]
+			encValEntry := &ethpb.Validator{}
+			if v, ok := s.validatorEntryCache.Get(key); ok {
+				valEntry, vType := v.(*ethpb.Validator)
+				if vType {
+					encValEntry = valEntry
```

Comment:
Is the validator retrieved from the cache copied or shared ? If its the latter you will need to perform a deep copy to prevent mutations.

Reply:
I am converting it to []bytes now, i think deep copy is not required after this change.